src/rgw/rgw_cache.cc

```diff
@@ -344,3 +359,424 @@ ObjectCache::~ObjectCache()
+  outstanding_write_size += len;
+  eviction_lock.Unlock();
+}
+
+bool DataCache::get(string oid) { 
+
+  bool exist = false;
+  string location = cct->_conf->rgw_datacache_persistent_path + oid; 
+  cache_lock.Lock();
+  map<string, ChunkDataInfo*>::iterator iter = cache_map.find(oid);
+  if (!(iter == cache_map.end())){
+    // check inside cache whether file exists or not!!!! then make exist true;
+    struct ChunkDataInfo *chdo = iter->second;
+    if(access(location.c_str(), F_OK ) != -1 ) { // file exists
+      exist = true;
+      { /*LRU*/
+
+	/*get ChunkDataInfo*/
+	eviction_lock.Lock();
+	lru_remove(chdo);
+	lru_insert_head(chdo);
+	eviction_lock.Unlock();
+      }
+    } else {	
+      cache_map.erase(oid);
+      lru_remove(chdo);
+      exist = false;
+    }
+  }
+  cache_lock.Unlock();
+  return exist;
+}
+
+size_t DataCache::random_eviction(){
+
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  ChunkDataInfo *del_entry;
+  string del_oid, location;
+
+  cache_lock.Lock();
+  n_entries = cache_map.size();
+  if (n_entries <= 0){
+    cache_lock.Unlock();
+    return -1;
+  }
+  srand (time(NULL));
+  random_index = rand()%n_entries;
+  map<string, ChunkDataInfo*>::iterator iter = cache_map.begin();
+  std::advance(iter, random_index);
+  del_oid = iter->first;
+  del_entry =  iter->second;
+  ldout(cct, 20) << "INFO::random_eviction index:"<< random_index << ", free size:0x" << std::hex << del_entry->size << dendl;
+  freed_size = del_entry->size;
+  free(del_entry);
+  del_entry = NULL;
+  cache_map.erase(del_oid); // oid
+  cache_lock.Unlock();
+
+  location = cct->_conf->rgw_datacache_persistent_path + del_oid; /*replace tmp with the correct path from config file*/
+  remove(location.c_str());
+  return freed_size;
+}
+
+size_t DataCache::lru_eviction(){
+
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  ChunkDataInfo *del_entry;
+  string del_oid, location;
+
+  eviction_lock.Lock();
+  del_entry = tail;
+  lru_remove(del_entry);
+  eviction_lock.Unlock();
+
+  cache_lock.Lock();
+  n_entries = cache_map.size();
+  if (n_entries <= 0){
+    cache_lock.Unlock();
+    return -1;
+    }
+  del_oid = del_entry->oid;
+  ldout(cct, 20) << "Engage1: lru_eviction: oid to remove" << del_oid << dendl;
+  map<string, ChunkDataInfo*>::iterator iter = cache_map.find(del_entry->oid);
+  if (iter != cache_map.end()) {
+    cache_map.erase(del_oid); // oid
+  }
+  cache_lock.Unlock();
+  freed_size = del_entry->size;
+  free(del_entry);
+  location = cct->_conf->rgw_datacache_persistent_path + del_oid; /*replace tmp with the correct path from config file*/
+  remove(location.c_str());
+  return freed_size;
+}
+
+
+void DataCache::remote_io(struct librados::L2CacheRequest *l2request ) {
+  ldout(cct, 20) << "Engage1: Add task to remote IO" << dendl;
+  tp->addTask(new HttpL2Request(l2request, cct));
+}
+
+
+std::vector<string> split(const std::string &s, char * delim) {
```

Comment:
1) we already have a str_split or something which works like this; 2) adam emerson wrote a better string splitter which doesn't create a stringstream

Reply:
removing, replacing with boost::split for now